contributivity.py

```diff
@@ -37,3 +93,1002 @@ def __str__(self):
+        # We are using this python implementation: https://github.com/susobhang70/shapley_value
+        # It requires coalitions to be ordered - see README of https://github.com/susobhang70/shapley_value
+        list_shapley_value = sv.main(nodes_count, characteristic_function)
+
+        # Return SV of each node
+        self.name = "Shapley values"
+        self.contributivity_scores = np.array(list_shapley_value)
+        self.scores_std = np.zeros(len(list_shapley_value))
+        self.normalized_scores = list_shapley_value / np.sum(list_shapley_value)
+        end = timer()
+        self.computation_time = end - start
+
+    # %% compute independent raw scores
+    def compute_independent_scores_raws(self, the_scenario):
+        start = timer()
+
+        print(
+            "\n# Launching computation of perf. scores of models trained independently on each node"
+        )
+
+        # Initialize a list of performance scores
+        performance_scores = []
+
+        # Train models independently on each node and append perf. score to list of perf. scores
+        for i in range(len(the_scenario.node_list)):
+            performance_scores.append(
+                self.not_twice_characteristic(np.array([i]), the_scenario)
+            )
+        self.name = "Independant scores raw"
+        self.contributivity_scores = np.array(performance_scores)
+        self.scores_std = np.zeros(len(performance_scores))
+        self.normalized_scores = performance_scores / np.sum(performance_scores)
+        end = timer()
+        self.computation_time = end - start
+
+    # %% compute independent additive scores
+
+    def compute_independent_scores_additive(self, the_scenario):
+        start = timer()
+
+        collaborative_score = self.not_twice_characteristic(
+            np.arange(len(the_scenario.node_list)), the_scenario
+        )
+
+        print(
+            "\n# Launching computation of perf. scores of models trained independently on each node"
+        )
+
+        # Initialize a list of performance scores
+        performance_scores = []
+
+        # Train models independently on each node and append perf. score to list of perf. scores
+        for i in range(len(the_scenario.node_list)):
+            performance_scores.append(
+                self.not_twice_characteristic(np.array([i]), the_scenario)
+            )
+
+        # Compute 'regularized' values of performance scores so that they are additive and their sum amount to the collaborative performance score obtained by the coalition of all players (nodes)
+        perf_scores_additive = softmax(performance_scores) * collaborative_score
+
+        self.name = "Independant scores additive"
+        self.contributivity_scores = np.array(perf_scores_additive)
+        self.scores_std = np.zeros(len(performance_scores))
+        self.normalized_scores = perf_scores_additive / np.sum(perf_scores_additive)
+        end = timer()
+        self.computation_time = end - start
```

Comment:
@Thomas-Galtier shouldn't we remove this as this is just a dilatation/contraction of the previous one?

Reply:
ok for me

Reply:
mince, du coup ça passe plus sur travis